rosbag2_cpp/test/rosbag2_cpp/test_sequential_writer.cpp

```diff
@@ -1108,6 +1108,103 @@ TEST_F(SequentialWriterTest, split_event_calls_on_writer_close)
   EXPECT_TRUE(opened_file.empty());
 }
 
+TEST_F(SequentialWriterTest, circular_logging_limits_number_of_files_by_max_bag_files)
+{
+  // Configure frequent splits and a small retention window
+  const uint64_t max_bagfile_size = 5;   // split every 5 writes
+  const uint64_t max_bag_files = 3;      // retain at most 3 files
+  const int message_count = 25;          // enough writes to exceed retention
+
+  auto sequential_writer = std::make_unique<rosbag2_cpp::writers::SequentialWriter>(
+    std::move(storage_factory_), converter_factory_, std::move(metadata_io_));
+  writer_ = std::make_unique<rosbag2_cpp::Writer>(std::move(sequential_writer));
+
+  auto message = std::make_shared<rosbag2_storage::SerializedBagMessage>();
+  message->topic_name = "test_topic";
+
+  storage_options_.max_cache_size = 0;             // direct writes
+  storage_options_.max_bagfile_size = max_bagfile_size;
+  storage_options_.max_bag_files = max_bag_files;  // enable bag file count circular limit
+
+  writer_->open(storage_options_, {"rmw_format", "rmw_format"});
+  writer_->create_topic({0u, "test_topic", "test_msgs/BasicTypes", "", {}, ""});
+
+  for (int i = 0; i < message_count; ++i) {
+    writer_->write(message);
+  }
+  writer_->close();
+
+  // After circular deletion, only max_bag_files files should remain in metadata
+  ASSERT_LE(fake_metadata_.files.size(), max_bag_files);
+  ASSERT_EQ(fake_metadata_.files.size(), fake_metadata_.relative_file_paths.size());
+}
+
+TEST_P(
+  ParametrizedTemporaryDirectoryFixture,
+  circular_logging_deletes_oldest_files_with_real_storage)
+{
+  // Integration test: verify max_bag_files actually deletes files from disk
+  const uint64_t max_bag_files = 3;
+  const size_t total_files_to_create = 6;
+  std::string topic_name = "topic";
+
+  rosbag2_storage::StorageOptions storage_options;
+  storage_options.uri = (fs::path(temporary_dir_path_) / "circular_bag").generic_string();
+  storage_options.storage_id = GetParam();
+  storage_options.max_bag_files = max_bag_files;
+
+  rosbag2_cpp::writers::SequentialWriter writer{};
+  writer.open(storage_options, rosbag2_cpp::ConverterOptions{});
+  writer.create_topic(
+  {
+    0u,
+    topic_name,
+    "test_msgs/msg/BasicTypes",
+    "cdr",
+    {},
+    ""
+  },
+  {
+    "test_msgs/msg/BasicTypes",
+    "ros2msg",
+    "bool bool_value",
+    ""
+  });
```

Comment:
Nitpick. A bit of simplification
```suggestion

  rosbag2_storage::TopicMetadata topic_metadata {
    0U, topic_name, "test_msgs/BasicTypes", "cdr", {}, ""
  };
  writer.create_topic(topic_metadata);```

Reply:
Updated.